File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.py
class Solution(object):
    def reverseParentheses(self, s):
        """
        :type s: str
        :rtype: str
        """
        open_parentheses = deque()
        result = []
        
        for curr_char in s:
            if curr_char == "(":
                open_parentheses.append(len(result))
            elif curr_char == ")":
                start = open_parentheses.pop()
                result[start:] = result[start:][::-1]
            else:
                result.append(curr_char)
                
        return "".join(result)

File: 1190-reverse-substrings-between-each-pair-of-parentheses/1190-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        stack<int> openParenthese;
        string result;
        for (char curChar:s){
            if(curChar == '(' ){
                openParenthese.push(result.length());   // store the current length as the start index for later and reversal

            } else if(curChar == ')' ){
                int start = openParenthese.top();
                openParenthese.pop();
                reverse(result.begin() + start, result.end());    // reverse the substring between the mathching parentheses

            } else {
                result += curChar;     // append non-parenthesis characters to the processed string

            }
        }
        
        return result;
        
    }
};